File: sdk-python/echoproxy/proxy.py
"""Drop-in proxy client for ``requests``.

Function signatures mirror ``requests.get/post/...`` so existing code only
needs the import swapped::

    import echoproxy.proxy as r
    r.get("https://api.example.com/v1/users")

Configuration is environment-driven::

    ECHOPROXY_API_KEY     required, the raw sk_live_... value
    ECHOPROXY_PROXY_URL   default http://localhost:8080
"""

from __future__ import annotations

import os
from typing import Any, Optional
from urllib.parse import urlparse

import requests
from requests.adapters import HTTPAdapter


class _SidAdapter(HTTPAdapter):
    """HTTPAdapter that rewrites the request URL to point at the proxy and
    adds the X-Echo-Key / X-Echo-Target headers. Mounted on both http:// and
    https:// schemes so any URL the caller passes is intercepted."""

    def __init__(self, proxy_url: str, api_key: str, **kw: Any) -> None:
        super().__init__(**kw)
        self.proxy_url = proxy_url.rstrip("/")
        self.api_key = api_key

    def send(self, request, **kw):  # type: ignore[override]
        u = urlparse(request.url)
        target = f"{u.scheme}://{u.netloc}"
        request.headers["X-Echo-Key"] = self.api_key
        request.headers["X-Echo-Target"] = target
        request.url = f"{self.proxy_url}{u.path or '/'}"
        if u.query:
            request.url += f"?{u.query}"
        return super().send(request, **kw)


def session(api_key: Optional[str] = None, proxy_url: Optional[str] = None) -> requests.Session:
    """Return a ``requests.Session`` that routes every call through the proxy.
    Pass explicit values or rely on environment defaults."""
    api_key = api_key or os.environ.get("ECHOPROXY_API_KEY", "")
    if not api_key:
        raise RuntimeError("echoproxy.proxy: ECHOPROXY_API_KEY env not set")
    proxy_url = proxy_url or os.environ.get("ECHOPROXY_PROXY_URL", "http://localhost:8080")
    s = requests.Session()
    adapter = _SidAdapter(proxy_url, api_key)
    s.mount("https://", adapter)
    s.mount("http://", adapter)
    return s


_default: Optional[requests.Session] = None


def _client() -> requests.Session:
    global _default
    if _default is None:
        _default = session()
    return _default


def configure(api_key: str, proxy_url: Optional[str] = None) -> None:
    """Override env-derived defaults. Useful in tests or multi-tenant apps."""
    global _default
    _default = session(api_key, proxy_url)


# ─── requests.* mirror ──────────────────────────────────────────────────────

def request(method: str, url: str, **kw: Any) -> requests.Response:
    return _client().request(method, url, **kw)


def get(url: str, **kw: Any) -> requests.Response:
    return _client().get(url, **kw)


def options(url: str, **kw: Any) -> requests.Response:
    return _client().options(url, **kw)


def head(url: str, **kw: Any) -> requests.Response:
    return _client().head(url, **kw)


def post(url: str, data: Any = None, json: Any = None, **kw: Any) -> requests.Response:
    return _client().post(url, data=data, json=json, **kw)


def put(url: str, data: Any = None, **kw: Any) -> requests.Response:
    return _client().put(url, data=data, **kw)


def patch(url: str, data: Any = None, **kw: Any) -> requests.Response:
    return _client().patch(url, data=data, **kw)


def delete(url: str, **kw: Any) -> requests.Response:
    return _client().delete(url, **kw)

File: sdk-ts/src/proxy.ts
/**
 * Drop-in replacement for the global `fetch` that routes every call through
 * the EchoProxy proxy. Same signature as `fetch`, so existing code only needs
 * the import swapped:
 *
 * ```ts
 * import { fetch } from "@echoproxy/sdk/proxy";
 * await fetch("https://api.example.com/v1/users");
 * ```
 *
 * Configuration is environment-driven and read at first use:
 *
 *   ECHOPROXY_API_KEY     required
 *   ECHOPROXY_PROXY_URL   default http://localhost:8080
 *
 * Override programmatically with {@link configure}.
 */

type ProxyConfig = {
  apiKey: string;
  proxyURL: string;
};

let _cfg: ProxyConfig | null = null;

function loadEnv(): ProxyConfig {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const env = (globalThis as any)?.process?.env ?? {};
  const apiKey = env.ECHOPROXY_API_KEY ?? "";
  const proxyURL = (env.ECHOPROXY_PROXY_URL ?? "http://localhost:8080").replace(/\/$/, "");
  return { apiKey, proxyURL };
}

function ensure(): ProxyConfig {
  if (_cfg === null) _cfg = loadEnv();
  if (!_cfg.apiKey) {
    throw new Error("echoproxy: ECHOPROXY_API_KEY env not set");
  }
  return _cfg;
}

/** Override env defaults. Useful in tests or multi-tenant apps. */
export function configure(apiKey: string, proxyURL = "http://localhost:8080"): void {
  _cfg = { apiKey, proxyURL: proxyURL.replace(/\/$/, "") };
}

/** Drop-in replacement for the global `fetch`. Same signature, same return. */
export async function fetch(input: string | URL | Request, init: RequestInit = {}): Promise<Response> {
  const { apiKey, proxyURL } = ensure();
  const targetURL =
    typeof input === "string" || input instanceof URL ? new URL(input.toString()) : new URL(input.url);
  const target = `${targetURL.protocol}//${targetURL.host}`;
  const proxied = `${proxyURL}${targetURL.pathname}${targetURL.search}`;
  const headers = new Headers(init.headers ?? (input instanceof Request ? input.headers : undefined));
  headers.set("X-Echo-Key", apiKey);
  headers.set("X-Echo-Target", target);

  // Preserve method/body when caller passed a Request.
  let finalInit: RequestInit = { ...init, headers };
  if (input instanceof Request && !init.method) finalInit.method = input.method;
  if (input instanceof Request && !init.body && input.body) finalInit.body = input.body;

  return globalThis.fetch(proxied, finalInit);
}

/* ─── convenience helpers — mirror common axios/got names ──────────────── */

export const get = (url: string, init?: RequestInit): Promise<Response> =>
  fetch(url, { ...init, method: "GET" });

export const head = (url: string, init?: RequestInit): Promise<Response> =>
  fetch(url, { ...init, method: "HEAD" });

export const del = (url: string, init?: RequestInit): Promise<Response> =>
  fetch(url, { ...init, method: "DELETE" });

export const post = (url: string, body?: unknown, init?: RequestInit): Promise<Response> =>
  send("POST", url, body, init);

export const put = (url: string, body?: unknown, init?: RequestInit): Promise<Response> =>
  send("PUT", url, body, init);

export const patch = (url: string, body?: unknown, init?: RequestInit): Promise<Response> =>
  send("PATCH", url, body, init);

function send(method: string, url: string, body?: unknown, init?: RequestInit): Promise<Response> {
  const headers = new Headers(init?.headers ?? {});
  let payload: BodyInit | null | undefined;
  if (body === undefined || body === null) {
    payload = body as null | undefined;
  } else if (typeof body === "string" || body instanceof FormData || body instanceof URLSearchParams || body instanceof ArrayBuffer || body instanceof Blob) {
    payload = body as BodyInit;
  } else {
    if (!headers.has("Content-Type")) headers.set("Content-Type", "application/json");
    payload = JSON.stringify(body);
  }
  return fetch(url, { ...init, method, headers, body: payload });
}
